analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionShouldNotBeThrownFromUnexpectedMethods.cs

```diff
@@ -85,61 +83,83 @@ protected override void Initialize(SonarAnalysisContext context)
+
+        private static void ReportOnInvalidThrowStatement(SyntaxNodeAnalysisContext analysisContext,
             SyntaxNode node, ImmutableArray<KnownType> allowedTypes)
         {
-            var throwToReportOn = node.DescendantNodes()
-                .OfType<ThrowStatementSyntax>()
-                .Where(tss => tss.Expression != null)
-                .Select(tss => new NodeAndSymbol(tss, analysisContext.SemanticModel.GetSymbolInfo(tss.Expression).Symbol))
-                .FirstOrDefault(tuple => tuple.Symbol != null &&
-                    !tuple.Symbol.ContainingType.IsAny(allowedTypes) &&
-                    !tuple.Symbol.ContainingType.DerivesFromAny(allowedTypes))
-                ?.Node;
-
-            if (throwToReportOn != null)
+            if (ExpressionBody(node) is { } expressionBody
+                && ThrowExpressionSyntaxWrapper.IsInstance(expressionBody.Expression)
+                && (ThrowExpressionSyntaxWrapper)expressionBody.Expression is { } throwExpression
+                && analysisContext.SemanticModel.GetSymbolInfo(throwExpression.Expression).Symbol is { } symbol
+                && ShouldReport(symbol.ContainingType, allowedTypes))
             {
-                analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(rule, throwToReportOn.GetLocation()));
+                analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, throwExpression.SyntaxNode.GetLocation(), "expression"));
             }
-        }
-
-        private static bool IsTrackedMethod(MethodDeclarationSyntax declaration, SemanticModel semanticModel)
-        {
-            var methodSymbol = semanticModel.GetDeclaredSymbol(declaration);
-            if (methodSymbol == null)
+            else if (node.DescendantNodes()
+                    .OfType<ThrowStatementSyntax>()
+                    .Where(x => x.Expression != null)
+                    .Select(x => new NodeAndSymbol(x, analysisContext.SemanticModel.GetSymbolInfo(x.Expression).Symbol))
+                    .FirstOrDefault(nodeAndSymbol => nodeAndSymbol.Symbol != null && ShouldReport(nodeAndSymbol.Symbol.ContainingType, allowedTypes)) is { } throwStatement)
             {
-                return false;
+                analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, throwStatement.Node.GetLocation(), "statement"));
             }
-
-            return methodSymbol.IsObjectEquals() ||
-                methodSymbol.IsObjectGetHashCode() ||
-                methodSymbol.IsObjectToString() ||
-                methodSymbol.IsIDisposableDispose() ||
-                methodSymbol.IsIEquatableEquals();
         }
 
-        private static bool IsTrackedOperator(OperatorDeclarationSyntax declaration)
-        {
-            return TrackedOperators.Contains(declaration.OperatorToken.Kind());
-        }
+        private static ArrowExpressionClauseSyntax ExpressionBody(SyntaxNode node) =>
```

Comment:
I think this method is generic enough and can be extracted outside this class to be reused by other rules.